constraint/pkg/client/client.go

```diff
@@ -1,355 +1,174 @@
 package client
 
 import (
-	"bytes"
 	"context"
-	"errors"
 	"fmt"
-	"path"
 	"sort"
 	"strings"
 	"sync"
 
-	"github.com/open-policy-agent/frameworks/constraint/pkg/apis/constraints"
+	apiconstraints "github.com/open-policy-agent/frameworks/constraint/pkg/apis/constraints"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/client/crds"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers"
-	"github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local"
 	clienterrors "github.com/open-policy-agent/frameworks/constraint/pkg/client/errors"
-	"github.com/open-policy-agent/frameworks/constraint/pkg/client/regolib"
-	constraintlib "github.com/open-policy-agent/frameworks/constraint/pkg/core/constraints"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/handler"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/types"
-	"github.com/open-policy-agent/opa/format"
 	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-type templateEntry struct {
-	template *templates.ConstraintTemplate
-	CRD      *apiextensions.CustomResourceDefinition
-	Targets  []string
-}
-
+// Client tracks ConstraintTemplates and Constraints for a set of Targets.
+// Allows validating reviews against Constraints.
+//
+// Threadsafe. Does not support concurrent mutation operations.
+//
+// Note that adding per-identifier locking would not fix this completely - the
+// thread for the first-sent call could be put to sleep while the second is
+// allowed to continue running. Thus, this problem can only safely be handled
+// by the caller.
 type Client struct {
-	driver  drivers.Driver
+	// driver contains the Rego runtime environments to run queries against.
+	// Does not require mutex locking as Driver is threadsafe.
+	driver drivers.Driver
+	// targets are the targets supported by this Client.
+	// Assumed to be constant after initialization.
 	targets map[string]handler.TargetHandler
 
-	// mtx guards access to both templates and constraints.
-	mtx       sync.RWMutex
-	templates map[templateKey]*templateEntry
-	// TODO: https://github.com/open-policy-agent/frameworks/issues/187
```

Comment:
From https://github.com/open-policy-agent/frameworks/issues/187
> Looks like we'll still need Constraints in Client. In fact, it will become the source of truth for Constraints.

why is constraints removed from Client?

Reply:
Constraints aren't removed from Client. They're now managed by `templateClient`, a subfield of `Client`.